docs/tutorials/using_datasets/IQL_torchrl.py

```diff
@@ -0,0 +1,522 @@
+actor = ProbabilisticActor(
+    module=actor_module,
+    in_keys=["loc", "scale"],
+    spec=action_spec,
+    distribution_class=TanhNormal,
+    distribution_kwargs={
+        "min": action_spec.space.low,
+        "max": action_spec.space.high,
+        "tanh_loc": False,
+    },
+    default_interaction_type=ExplorationType.MODE,
+)
+
+# %%
+# For convenience, we gather the actor and value functions into a single "model":
+
+# %%
+model = torch.nn.ModuleList([actor, qvalue, value_net]).to(device)
+
+# %%
+# Under the hood, ``MLP()`` uses `LazyLinear <https://pytorch.org/docs/stable/generated/torch.nn.LazyLinear.html>`_ layers, whose shape is inferred during the first pass. Later methods need a fixed shape, so we forward some random data through the network to initialise the Lazy modules:
+
+# %%
+with torch.no_grad(), set_exploration_type(ExplorationType.RANDOM):
+    tensordict = env.reset().to(device)
+
+    for net in model:
+        net(tensordict)
+
+# %%
+# Loss and optimizer
+# ~~~~~~~~~~~~~~~~~~
+
+# %%
+# The explicit details of the Implicit Q-Learning losses are captured by TorchRL's ``IQLLoss`` module:
+
+# %%
+loss_module = IQLLoss(
+    model[0],
+    model[1],
+    value_network=model[2],
+    loss_function="l2",
+    temperature=3,
+    expectile=0.7,
+)
+loss_module.make_value_estimator(gamma=0.99)
+
+# %%
+# IQL uses "soft updates" for the target network :math:`Q_{\hat{\theta}}(s, a)`. The target network parameters are slowly updated in the direction of :math:`Q_{\theta}(s, a)` at each iteration via Polyak averaging :math:`\hat{\theta} \leftarrow \tau \,\theta + (1 - \tau) \, \hat{\theta}`,
+
+# %%
+target_net_updater = SoftUpdate(loss_module, tau=0.005)
+
+# %%
+# We optimize all of the networks using Adam:
+
+# %%
+optimizer = torch.optim.Adam(loss_module.parameters(), lr=0.0003)
+
+
+# %%
+# Training
+# ~~~~~~~~
+
+# %%
+# To demonstrate training, we run IQL for 50,000 iterations. During training, we will evaluate the policy every 1000 iterations. But note that this is for evaluation purposes only. Unlike online RL, we do not collect new data during training.
+
+# %%
+@torch.no_grad()
+def evaluate_policy(env, policy, num_eval_episodes=20):
+    """Calculate the mean cumulative reward over multiple episodes."""
+    episode_rewards = []
+
+    for _ in range(num_eval_episodes):
+        eval_td = env.rollout(max_steps=max_episode_steps, policy=policy, auto_cast_to_device=True)
+        episode_rewards.append(eval_td["next", "reward"].sum().item())
+
+    return np.mean(episode_rewards)
+
+
+# %%
+# The training loop is essentially the standard PyTorch gradient descent loop:
+#
+# 1. Sample a batch of transitions from the dataset :math:`\mathcal{D}`.
+# 2. Compute the loss :math:`\ell = L_V(\psi) + L_Q(\theta) + L_\pi(\phi)`.
+# 3. Backpropagate the gradients and update the networks, including the target Q-network.
+
+# %%
+from tqdm.auto import tqdm
+
+iterations = 50_000
```

Comment:
the loss plot seems to converge after a few iterations (the dataset is very small)
Also, the return is not clear if it is benefiting from the "long" training.

Maybe we can reduce the number of iterations or use the "pen-expert-v1" which is larger

Reply:
To keep things fast, it only uses a small number of rollouts (20) to estimate the return, so the return plot is rather noisy. I haven't tested it in detail but 50K iterations seems to be the amount required to get a decent return across a range of seeds. However I decreased the number of iterations to 10K so that it doesn't timeout the tests.